### DISCUSSION TITLE
Capsules dependency management

### DISCUSSION THREAD

**Discussion by antonioribeiro** [area17/twill#979] [General]
Hello everyone!

Not sure if any of you are using Capsules already, but I'm finishing a project at AREA 17 with them (at least one other internal Team is also using it) and it feels good to use it so far. If you need help using it, just mention me here.

Well the next step would be to improve the installer, which exists already but it's not totally ready for prime time yet. I'm working (yes WIP) on 3 different public Capsules now to better understand the needs, and @ifox#4260 came to me one of these days thinking about the --copy option not being the best one for Capsules. So now I'm second guessing @ifox#4260's second guess :smile:

For everyone to better understand the what we are thinking, we have two installing options:

- Installing with --copy (implemented, but miss a lot of things)
- Installing with --require (not implemented)

Currently the **COPY** command just downloads the zip and unzip it on your /Capsules directory. Implementation is very rudimentary, but it's gettind the job done. But I believe we still need it to do more:

- Have a dependency management (the public Homepages Capsule depends on Base, as Cities Capsule depends on Countries and Base). A PR was sent and this is also working fairly good. I did a PR for this: https://github.com/area17/twill/pull/933.
- A way to rename the Capsule while installing it: sometimes we need the same behaviour with a different name, but Capsules (modules) contains a lot of classes and files, manually renaming them now is easier because everything is on a single directory, but it still take some time and effort.
- We won't get any updates for the Capsule, is this a problem?

The **REQUIRE** command would, in other hand, use Composer as a dependency manager, which is, of course, fantastic, but we have some things to think about here: 

- Are we installing Capsules on the vendor folder?
- We can probably tell Composer to put them on a app/Twill/Capsules, as, I believe, Drupal does.
- And if I want to make changes to the Capsule, should I be forced to fork it on Github and change something small only for that particular project?
- Or should I just delete the .git repository, remove it from the composer.json and commit it definitely on my project?
- If the Capsules maintainer changes something on the Capsule, my project can break, right? Should we care about this? This is how package dependency system works after all.
- Do we really need Capsules to be upgradable? Why?
- A Capsule that has a composer.json, can have all sort of dependencies (packages and other Capsules). We probably want packages to get updates, but we still want Capsules to receive changes?

I know I'm asking the same question over and over, using different words, but I feel that Capsules are so connected to the project we are building now, because as soon as you install a Capsule you may want to add or remove fields from it, or change the order of the fields in the form, or just disable translations because your project has only one language. So I feel that they are too much of a living being and cannot do everything, for everyone, so that "require" option could pose more problems than solutions.

Unless we keep the dependency management for packages and disable it for the Capsule itself and just let users to modify them if they want.

What do you think about all this? 

Cheers!

**Comment by pboivin** [comment]
Hey @antonioribeiro!

A lot to think about here :)

I haven't started using Capsules yet, but I really see the appeal!

I think you put the problem in context quite well. From my point of view, the first question that comes to mind reading your message is: _Are Capsules intended as "boilerplate" to be customized and extended, or more as "plugins" to be integrated and built around?_

At a glance, it looks like `--copy` supports the "boilerplate" approach, whereas `--require` supports the "plugin" approach. Does that make any sense? I think that both use cases are likely to happen in the wild.

I agree that dependency management is a must. For what it's worth, I also think that a line in a README that says "to use this Capsule, you must run `composer require xyz`" is pretty okay as a form of basic dependency management...

I really see that the question of why and how to update a Capsule is at the heart of the problem. I don't see how Capsules as "boilerplate" can be updated — once it's in your project, it's yours. On the other hand, Capsules as "plugins" can very well be updated, whether via composer or just unzipping, really.

Not sure if I've actually added new thoughts to the discussion at this point but, I'm interested in the subject ;)

**Comment by kirkbushell** [comment]
Using composer to manage the capsules and a dependencies is imho, a no-brainer. It offers a wealth of opportunities and advantages over manually doing this. That said, publicly available capsules should be configurable just like the rest of the system, rather than being copied or "published" to your local project (I see this introducing a world of problems, the main one being, as you suggested - capsule updates).

They should be viewed as mini repositories with a dependency on Twill versions.

**Reply by pboivin** [reply]
@kirkbushell I agree with you on the model of Capsules as mini repositories. Although, I also see how — a Capsule being made of "Twill stuff" (migrations, models, views, ...) — there could be a point where configuration options just aren't enough. As @antonioribeiro mentioned, when working with a Capsule, I may want to reorder fields, add some extra processing over here, update the styling over there, etc. At some point in this process, it's just easier to `--copy` or fork than to configure. Does that make any sense?

**Comment by ibes** [comment]
My background:
Haven’t used Capsules. Coming from WordPress. 

I see use cases for both require and copy. 

But first of all:
Telling composer to install certain packages in certain places is not difficult. The main composer.json needs to know the rules and the packages should have a certain type to be recognizable for the rule. 
I know that from WP with composer setups. 

In WP there is the rule to never touch a plugin directly to not break my own modifications by updating the plugin. 
To still be able to modify the whole hook system is needed. Filters, Actions.
That works ok. 

I see a use case for --require mostly in high optimized or specialized stuff. 
SEO or media handling. Where I want to get delivered with the hottest updates.

I see --copy more where I want a boilerplate with best practice that I can quickly delete down to what I need and customize.
A “customer feedback” content type, a blog setup, a frontend auth handling 


Maybe capsules need to be split in two things?
Maybe it is cool to just give people both options so they can go for what works best. 


Could there be a mix of boilerplate and updatable parts. 
Having the most important parts in a core capsule that is expended by the boilterplate capsule. 
Is that the “publish” mechanism of Laravel?!

I guess for trivial stuff I would just want to get a head start.
For difficult things I would want to get the most difficult/complex things updated while being able to adjust it to my setup.

**Comment by antonioribeiro** [comment]
Thanks for your feedback @pboi20 , @kirkbushell , @ibes!

Looks like **REQUIRE** is a must, so let's try to do this, and see how it plays. I believe it could work well for really big, super complex, or highly configurable Capsules, where people would not feel the need to change as soon as they touch it.

@ibes , you came up with a third kind of capsule: **HYBRID**, nice one! But the way Twill modules (and Laravel) are made, a very simple update (like adding a new field) require changes to at least migration and model, if you are adding a relationship, you may have to change the repository, if the index or browser are customized, you may also have to change the controller. And I'm just oversimplifying it, because Twill is huge and pretty much anything could require a change to a class, which would prevent the Capsule from receive updates. But I would love to hear more about what could be the implementation for this.

Maybe we can, in the future, make Capsules less dependable of some classes, relying more in configuration (I'm more a configuration over convention guy), using DTOs or configuration objects (as pointed by @kirkbushell on #internals @ our Discord), but I feel we have too much unneeded boilerplate code. For example, seems to me that we don't really need this kind of controller, and I have many of those:

``` php
<?php

namespace App\Twill\Capsules\Collections\Http\Controllers;

use A17\Twill\Http\Controllers\Admin\ModuleController;

class CollectionController extends ModuleController
{
    protected $moduleName = 'collections';

    protected $indexColumns = [
        'title' => [
            'title' => 'Title',
            'field' => 'title',
        ],

        'city' => [
            'title' => 'City',
            'field' => 'city_name',
        ],
    ];
}
```

We can probably have a configuration for the index and a more global Controller class handling the requests. Same for migrations,  we could probably use an object like Laravel's validation `Rule` to define migrations, same for Repositories (if we are only adding `afterSave()` and `getFormFields()` to handle browsers, repeaters, relationships, but models in the other hand looks harder.

Laravel Nova also uses it extensively, that's probably the cleverest part of it, and it's also super powerful:

``` php
public function fields(NovaRequest $request)
{
    return [
        Text::make('First Name'),
        Text::make('Last Name'),
        Text::make('Job Title'),
    ];
}
```

We could possibly do forms this way, like Nova too. But Twill forms are also a different kind of beast.
 
If we are able to do something like this, maybe we can have **HYBRID** Capsules, configurable and upgradable. But this is how I see it.

What about you?

**Reply by kirkbushell** [reply]
The only difference between Twill forms and Nova forms (other than configuration) is that Twill has the block editor (blockable) fields. As long as that is taken into account when developing the solution, it would be ideal.

The way I see it currently, is that if I could use the block editor in Nova, that would be my go-to, every single time. Nova does everything better, except content. Twill imho is a leader in that space, and combining the projects or even supporting Nova through a paid Twill package (I would HAPPILY pay for a block editor plugin for Nova!!!)

**Reply by antonioribeiro** [reply]
I can tell more things Twill does a lot better than Nova: 

- Image/File management
- Image cropping: this is something I didn't see anywhere else, not the way it was implemented on Twill.
- In form translations (does Nova even have translations?)
- Preview: can you preview your frontend inside Nova, and view pages exactly the way your users will see them?
- Revisions: can you visually compare (preview again) two pages and choose to rollback to an older one?
- Easily publish/unpublish records. Maybe we can have this with some plugin in Nova, adding it automatically to all forms, but Twill's implementation is also beautiful and has a great UX, and that counts too, right? 😜

**Reply by kirkbushell** [reply]
Yup, all fair points :)

In all honesty though - the rest to me doesn't matter. Just give me the content block editor. lol

Revisions are actually really nice, I reckon that could be written as a plugin as well. Translations are a much more difficult problem to solve.

Also - Nova isn't mobile-friendly at all.

**Reply by slimko** [reply]
@kirkbushell free  block editor for Nova https://github.com/whitecube/nova-flexible-content

**Reply by kirkbushell** [reply]
> @kirkbushell free block editor for Nova https://github.com/whitecube/nova-flexible-content

I've ended up using this now in several projects. Thank you :)

**Comment by antonioribeiro** [comment]
Hello everyone, my first take on making Composer install Capsules into our own Capsules folder failed, because I failed to understand the concept of a `project` in Composer.

As most of you know, Composer allows frameworks to configure [installer-paths](https://getcomposer.org/doc/faqs/how-do-i-install-a-package-to-a-custom-path-for-my-framework.md), what I didn't know when [I first tried this](https://github.com/composer/composer/issues/10090) is that this feature is tied to a `project`, so what's the project? It's the main Framework on what your application is based on. In our case this Framework is Laravel, not the Laravel Framework package, but [laravel/laravel](https://github.com/laravel/laravel/blob/8.x/composer.json#L3). In the case of Drupal and Wordpress, it's of course Drupal and Wordpress main repos, and no a third party package like Twill. There a [list of the supported ones here](https://github.com/composer/installers).

I believe (yes I'm still not totally sure) that in order for this to work our application `composer.json` would have to be based on a "project" `composer.json` which should look something like this:

``` JSON
{
    "name": "twill/project",
    "type": "project",
    "require": {
        "php": "^8.0",
        "area17/twill": "^2.5.1",
        "area17/twill-capsule-homepages": "^1.0",
        "laravel/framework": "^8.54"
    },
    "extra": {
        "installer-paths": {
            "app/Twill/Capsules/Vendor/{$name}": ["type:twill-capsule"]
        }
    }
}
```

This is not impossible, of course, but this means a lot of changes to the Twill ecosystem, like creating a "Twill Project" repository, based on the laravel/laravel, of course, possibly an installer and who knows what more...

Any ideas on how could we tackle this without going down this path?

Cheers!

**Reply by kirkbushell** [reply]
I think there's one question that needs to be asked and that is - why do it this way? Why not just allow composer defaults, developers install a package as per normal? I don't see the benefit of having to have capsules in their own folder.

Developers just need the appropriate hooks to be able to install their capsules.

**Reply by pboivin** [reply]
@kirkbushell If I understand correctly, what you're describing boils down to:

```
composer require area17/twill-capsule-homepages
```

which installs under `vendor/` like a regular composer package. Then you add the capsule to `config/twill.php` and are good to go. Or am I missing something?

**Reply by kirkbushell** [reply]
Yup, pretty much - or have some sort of auto-discovery feature.

**Reply by ifox** [reply]
I think I agree with @kirkbushell here. @antonioribeiro what is the goal of installing those dependencies outside of the vendor folder? This actually feels like going backwards to me, no? I really don't see what a regular Laravel package can't do inside of a Twill application. They have all the features we need to publish and configure files as needed, so why would we go down the path that Drupal and WordPress took?

**Reply by ifox** [reply]
Rereading myself it sounds accusatory, when you're only doing some research, my bad! Would love to hear your take on regular Laravel packages and their limitations in the context of Twill.

**Reply by kirkbushell** [reply]
I want to expand on this actually - I don't understand the need for capsules, at all. It sounds like a solution to a problem that doesn't exist, but like ifox - perhaps if I understood the business case better I could offer a more constructive response.

**Reply by ifox** [reply]
There is definitely a business case for capsules @kirkbushell, however they end up being implemented, the goal is to create reusable Twill modules that can be shared with the community, both for demonstrations purposes, but first and foremost for consistency and quality. As an agency working with many clients, we are often solving the same problems with very similar content structure, when we'd rather all work on a system that can flex depending on needs. Capsules should enable that, and they should also be a way to bootstrap a Twill project very quickly. Need a basic blog, get the posts capsules, need recurrent events, get the recurring events capsule, and so on...

**Reply by ifox** [reply]
They are also, simple as that, a different folder structure to work with Twill modules. It's all about preference at this stage as modules and capsules are the same thing organized differently.

**Reply by kirkbushell** [reply]
If it's just about standards, I would advise then the following:

1. A standard approach to module development, codified
2. An artisan command to setup the structure in a new project

This would allow us to use existing solutions and just enable the creation of new "capsules" very easily without having to reinvent the wheel.

**Reply by ifox** [reply]
It's not about standards only, it's about content structures and reusable modules configuration. Say you have a very complex events module, with a lot of form fields and custom logic, how do you make that reusable in another Twill app without copying all the files a module is made of? My point of view is that a regular Laravel package would do, but I think @antonioribeiro has counter arguments to that.

**Reply by kirkbushell** [reply]
I think that is a problem with Twill's management of configuration, which makes it more difficult to manage complex, custom setups.

In either case, a decent plugin architecture shouldn't require a particular format (although that certainly has benefits in the broader community), but it shouldn't be necessary to setup and register a new capsule.

Imho, we should be able to simply register a service provider that tells Twill where everything it needs, is.

**Reply by kirkbushell** [reply]
In the boot method of a service provider, being able to do something like: Capsule::register(new MyCapsule) should be sufficient.

**Reply by antonioribeiro** [reply]
The more I use Capsules, the more I feel lost going back to old projects and having to deal with modules files dispersed all over the application folder structure. It's still the same Twill, it's just more organized. One simple task that drives me crazy: try deleting a module from your application, I can tell you it's not an easy task as I have been bitten by it once or twice. Having a  namespace exclusive for the Capsule, like `App\Twill\Capsules\Countries` makes it easier to find all classes and configuration files referencing it. In the traditional modules way, we need to search for `App\Models\Country`, then `App\Repositories\CountriesRepository` (see the difference between singular and plural classes? yeah, that too), and so on.

Of course, this is something that may only only make sense for companies building lots of similar applications (we have many application built for museums, for example) at high speed and requirements changing during development, so maybe you could ask yourself: "How can someone need to delete a whole module from an application?! I can't see an use case for something like this". So as a package developer and contributor I understand that's it's frequently hard to see usefulness on something other people think "no one" would need, use cases can be very particular but should not be disregarded or overlooked, just because we, particularly, don't see the point.

We have some ideas here:

### Composer install - The Frameworks Way
Not sure if you understood, but what I was try to build also boils down to:

```
composer require area17/twill-capsule-homepages
```

The difference is that instead of having the Capsule installed on 

```
vendor/area17/twill-capsule-homepages
````

It would put the Capsule at

```
App/Twill/Capsules/Vendor/A17/Homepages
```

And Composer would just keep updating the capsule at every `composer update`.

But this way if I need to tweak anything on the Capsule, because of application requirements, I would have two options:

1) Fork it, change, commit, push and just use my fork. 

This can also be done on /vendor, of course, but we could also 

2) Delete the capsule from my `composer.json` and just commit it into the project, and BOOM, done, nothing else has to be done.

Having it on the `vendor` folder works as well, just make this a bit harder if I just want to commit the capsule directly to my application folder.

**Verdict**
Installing them on the `App\Twill\Capsules` folder using Composer seems hard right now. But as I said before, at least for our use cases, where design are crafted strategically for each client, which means every single page is deeply thoroughly rethought for every new museum (or any other client) we work for. A very simple page like a landing Homepage has probably completely different structures on all of our clients, event if it's a single input different. So, at least for me, now, silver-bullet Capsules are utopic, hopefully I'll be wrong tomorrow, unless, of course we start to create highly configurable Capsules, which, also, are hard to build and maintain. But we have the `--copy` option we already have install them, so we can just keep it.

### Twill capsule:install --copy
(TL;DR: this one it's just for people that didn't get our use case yet). As I said before, at least for our use cases, where design are crafted strategically for each client, which means every single page is deeply thoroughly rethought for every new museum (or any other client) we work for. A very simple page like a landing Homepage has probably completely different structures on all of our clients, event if it's a single input different. So, at least for me, now, silver-bullet Capsules are utopic, hopefully I'll be wrong tomorrow, unless, of course we start to create highly configurable Capsules, which, also, are hard to build and maintain. But we have the --copy option already done to install them, so we can just keep it.

### Just use the vendor folder
Install them as usual Composer packages, let them live on the `vendor` folder. We can probably do this kind of easily, we just need to, find a way to make them announce themselves to Twill, like Laravel package discovery. Meaning one more command to tell developers to add to their `post-autoload-dump` on `composer.json`, or we could maybe use the `twill:update` command to do discovery too...? But I believe we have some options here:

- Another `twill:package:discover` command
- Use it on `twill:update`
- Just add the Capsule vendor path (or whatever path) to `config('twill.capsules.homepages')`.

### Plugins
Capsules are not meant to remotely resemble to a plugin. This is another feature to be added to Twill, so let's maybe break them down:

#### Capsules / Modules
Meant to control data structures and business requirements. So if you need to create a event system on your application, Capsules will: 

- Manage the database tables and columns.
- Have the CMS CRUD module with forms, blocks, media, orders, filters, browsers.
- Manage and enforce business logic.

#### Plugins
Pluggable code that can be use to change Twill's behaviour or add new features to Twill, a few examples:

- Integrated translation system, where client admins can tweak translation files directly on Twill instead of us developers doing it on `resource/lang/en/messages.php` kind of files.
- Current logged in users in the application
- AWS, Azure, Google Cloud services watchers
- Forge or Vapor panel.
- Horizon (for Queues watch)
- A full application health panel (at some point I will try to integrate [this one](https://github.com/antonioribeiro/health) to Twill).

**Reply by kirkbushell** [reply]
I'm just gonna put it out there - I am not a fan, of having 3rd party capsules be installed to a local location. Honestly, I don't even see the point. This feels like a backwards step in regards to a properly supported module system.

I totally get and acknowledge wanting to define standard structures that the community can follow, and that could be easily supported with new twill commands. Fully behind that and support it. I think expanding on the twill module system to support this better, is a good move.

But this proposal feels like a step backwards, it sounds messy and archaic, just so people can copy-paste code into their projects (which they could do if they wanted anyway).

What is wrong with a solid plugin architecture?

I guess what I'm getting at here is, let's not mix and match requirements for modules between custom setups and community-provided capsules. They should both function in the same way, but can easily leverage existing package management solutions. One just lives in vendor, the other in your app directly.

Tbh though, it shouldn't matter WHERE they live.

**Reply by pboivin** [reply]
@kirkbushell  I agree that `composer install`-ing anywhere other than `vendor/` feels a bit strange. It's something other projects do, so it's not exactly uncommon... I remember working on elaborate wordpress setups where composer packages landed on `vendor/` but plugins landed on `app/plugins/`, with a mix of committed and git-ignored items. Not sure I can thoroughly explain why, but it certainly felt half baked.

**Reply by pboivin** [reply]
@antonioribeiro  I appreciate all the thoughts you've been putting into this problem! This all makes sense to me, especially the need to `--copy` for bespoke applications.

If I may just share a minor nitpick:

I disagree on the need to differentiate between plugins and capsules. I think there's a lot of overlap conceptually, and the distinction is really mostly philosophical. Say I'm developing a "plugin" that needs to hook into Twill internals as well as: install views, models, repositories, a CMS menu entry... What if I find a way to wrap that in a capsule as a means of distribution?

I guess what I'm saying is mostly — We don't need to introduce a "plugin" entity into the mix. I believe vanilla laravel packages already fit the bill. We may just need more (a lot more) "hooks" to extend what is possible.

**Reply by kirkbushell** [reply]
@pboivin Actually there is a point of difference between the two. I would think that all capsules are plugins, but not all plugins are capsules. Ie. Maybe a plugin allows you to change how user registrations are done (socal auth, perhaps?), but a capsule is a full CRUD "plugin" installed, maybe like a comments capsule?

There is certainly a difference between the two, but both should behave from a dependency perspective in the same way.

**Reply by pboivin** [reply]
@kirkbushell I fully agree and I believe we are arriving at the same conclusion – Capsules are a kind of plugin and nothing new is needed to support other kinds of plugins. Mostly "hooks" to extend Twill core.

**Reply by kirkbushell** [reply]
> @kirkbushell I fully agree and I believe we are arriving at the same conclusion – Capsules are a kind of plugin and nothing new is needed to support other kinds of plugins.

Quite the contrary, quite a lot is needed to support a decent plugin architecture for Twill :) But I believe Capsules should be an extension of that system, and is really where the work should begin.

**Reply by freshleafmedia** [reply]
While it doesn't address the plugin problem I've had a crack at making Capsules Composer installable: #1291

They can co-exist with Capsules in the `app/Twill/Capsules` directory so this allows users to have their own bespoke Capsules while also including common Capsules from Packagist

**Comment by antonioribeiro** [comment]
I know we can do anything with code, the question we have to pose ourselves is: should we?

Just to scratch the surface of a plugin system for Twill here, I believe a real plugin system for Twill would have to have support for VueJS. So are we adding JavasScript compilation to Modules now? Because this is clearly something that Modules doesn't support today, and, as said before Capsules are nothing more than Modules living on their own directory tree. And if we don't do this, well, Capsules will also start to become incompatible with Modules, or maybe just do more things than Modules do and, but this can be very confusing.

**Reply by kirkbushell** [reply]
Yup, and Vue JS doesn't play very nicely at all with a decent plugin architecture - one of it's biggest flaws.

Laravel Nova handles this rather well, imho. There could be some lessons there that we could take onboard.
